.circleci/config.yml

```diff
@@ -0,0 +1,115 @@
+#   https://circleci.com/docs/2.0/configuration-reference
+####################
+# CircleCI built-in environment variables:
+#   https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
+####################
+
+
+####################
+# Templates: see "anchors" in https://learnxinyminutes.com/docs/yaml/
+####################
+
+save_cache_settings: &save_cache_settings
+  key: python_mozetl-{{ .Branch }}-{{ .Environment.CIRCLE_WORKFLOW_ID }}
+  paths:
+    - ~/python_mozetl/.tox
+
+prep_cache_settings: &prep_cache_settings
+  name: Clean directories to restore
+  command: |
+    : # no-op
+
+restore_cache_settings: &restore_cache_settings
+  keys:
+    - python_mozetl-{{ .Branch }}-{{ .Environment.CIRCLE_WORKFLOW_ID }}
+    - python_mozetl-{{ .Branch }}
+    - python_mozetl-
+
+common_settings: &common_settings
+  docker:
+    - image: python:2.7
+  working_directory: ~/python_mozetl
+
+# CircleCI generally only triggers builds on commits to the main repository,
+# so PRs coming from branches of the main repo simply reference builds for existing commits
+# (CIRCLE_BRANCH=branchname and CIRCLE_PR_NUMBER is unset);
+# The only time PRs will trigger a build is when the PR is referencing a fork
+# (CIRCLE_BRANCH=pull/XXX and CIRCLE_PR_NUMBER=XXX).
+early_return_for_forked_pull_requests: &early_return_for_forked_pull_requests
+  name: Early return if this build is from a forked PR
+  command: |
+    if [ -n "$CIRCLE_PR_NUMBER" ]; then
+      echo "Nothing to upload for forked PRs, so marking this step successful"
+      circleci step halt
+    fi
+
+early_return_for_skip_tests: &early_return_for_skip_tests
+  name: Early return if the latest non-merge commit message contains "[skip-tests]"
+  command: |
+    COMMIT_MESSAGE=$(git log --format=%B --no-merges -n 1)
+    if [[ "$COMMIT_MESSAGE" =~ "[skip-tests]" ]]; then
+      echo "Skipping tests due to [skip-tests] flag, so marking this step successful"
+      circleci step halt
+    fi
+
+
+####################
+# Jobs: see https://circleci.com/docs/2.0/jobs-steps/
+####################
+
+version: 2
+jobs:
+
+  test:
+    <<: *common_settings
+    parallelism: 8
+    steps:
+      - checkout
+      - run: *early_return_for_skip_tests
+      - run: *prep_cache_settings
+      - restore_cache: *restore_cache_settings
+      - run: apt update && apt install -y libsnappy-dev default-jre
+      - run: pip install tox
+      - run:
+          name: run tests
+          command: |
+            circleci tests glob "tests/**/test*.py" > tests.txt
+            tox -- \
+              $(circleci tests split --split-by=timings tests.txt | tr -d '\r') \
+              --junitxml=test-reports/junit.xml
+      - store_test_results:
+          path: test-reports
+      - store_artifacts:
+          path: test-reports
+      - run:
+          name: Submit code coverage data
+          command: |
+            bash <(curl -s https://codecov.io/bash)
+      - save_cache: *save_cache_settings
```

Comment:
Maybe we should put the save_cache step under lint. CircleCI relies on the saved caches being immutable, so they must always have unique names, but this as written will attempt to save an identical cache on each of the 8 parallel test jobs.

Since the lint job isn't parallel, it's an easier home for saving the cache.

Reply:
This hasn't caused any issues, but I can see how overwriting the same file 8 times is less than ideal.

Reply:
There's also a an option for a `deploy` step (I think that's what it's called) that only runs for the index=0 job.